Separate interface for solving under assumptions [depends: 4450, blocks: 4054]

peterschrammel opened the pull request:
Based on #4450, only review last commit.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

peterschrammel committed:
Rename prop_convt to decision_procedure_incrementalt

decision_procedure_incrementalt will then be further split
into support for assumptions and contexts.

peterschrammel committed:
Update decision procedure docs

peterschrammel committed:
Split out decision procedure with solving under assumptions

Makes it explicit which algorithms actually require this feature.

martin-cs commented on src/solvers/prop/decision_procedure_assumptions.cpp:
```diff
@@ -0,0 +1,20 @@
+
+void decision_procedure_assumptionst::set_assumptions(const bvt &)
+{
+  UNREACHABLE;
```
Could this be pure virtual or something else?  Base implementations of methods that are just "nope!" make me kinda nervous.

peterschrammel replied:
True, I've actually thought I had patched that. Will change.

peterschrammel replied:
Done

martin-cs commented on src/solvers/prop/decision_procedure_assumptions.h:
```diff
@@ -0,0 +1,29 @@
+{
+public:
+  /// Set assumptions for the next call to operator() to solve the problem
+  virtual void set_assumptions(const bvt &);
```
Why is this / should this be a separate call?  If the solver is deterministic then it's not like you are going to call solve multiple times?  One reason it might be separate is that you want to be able to call this multiple times, but, if that is the case, do you need interfaces to manage and remove these?

peterschrammel replied:
There is no particular reason except legacy. You can remove assumptions by passing an empty `bvt`.

martin-cs replied:
If we are going to refactor then really the assumptions should be an argument to the solve call and not stateful.

martin-cs replied:
Better yet; it can be an argument and a result of the solve call so we can get back a reduced set of assumptions that are sufficient to give UNSAT.

martin-cs reviewed (commented):
Am unsure but I guess it makes sense to resolve the discussion around #4450 first.

peterschrammel commented:
@martin-cs, another question is whether `is_in_conflict` should be placed on top of `assumptions`. This would be justified as `smt2_convt` can emulate assumptions, but cannot do `is_in_conflict`.

peterschrammel committed:
Split out is_in_conflict solver capability

This is only provided by the prop_conv_solvert-based hierarchy
at the moment and is quite specific to MiniSAT-based solvers.
The functionality itself is used out-of-tree only (2LS).

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: a17fb3696880178346e020a6dfa7104daff4e034).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106239885
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

martin-cs commented:
To my mind `is_in_conflict` is something for an incremental solver interface (i.e. the main interface) rather than solve under assumption.  If you are monotonically adding to a set of constraints then tracking (maybe triggering solve?) when a call to solve has returned UNSAT is a meaningful thing to do.  If you are working with assumptions then it is just making the interface more stateful than it needs to be.

peterschrammel commented:
@martin-cs, no. Minisat Solver.cc:401: `...express the final conflict in terms of assumptions. Calculates the (possibly empty) set of assumptions that led to the assignment of 'p', and stores the result in 'out_conflict'.`

martin-cs commented:
@peterschrammel : my point is that this could be wrapped by this API and made into a single call to `solve : set of assumptions -> SAT | set of contradictory assumptions`.  The actual solver is going to be very stateful but our interface to it can be a lot less stateful.

peterschrammel commented:
obsolete
